### PR TITLE
[RUM-11388] Fix FileBasedConfiguration related issues

### PR DESCRIPTION
**Pull request by sbarrio**
### What does this PR do?

This PR aims to fix several issues with FileBasedConfiguration initialization of the SDK.

1.- Fixes a [reported issue](https://github.com/DataDog/dd-sdk-reactnative/issues/960) that makes it so directly passing a configuration json file enforces users to duplicate and nest another `configuration` property for it to work, ignoring the structure defined on `datadog-configuration.schema.json`. Doing so makes the schema validator label the json configuration as invalid. I fixed it so it properly accepts the schema we enforce with just one `configuration` entry.

This means that creating a new FileBasedConfiguration can be done like so:

```ts
// 1.- File based configuration from .json
const configuration = new FileBasedConfiguration(require("../datadog-configuration.json"));

// 2.- File based configuration from .json and custom mapper setup
const configuration = new FileBasedConfiguration( {
   configuration: require("../datadog-configuration.json").configuration, 
   errorEventMapper: (event) => event, 
   resourceEventMapper: (event) => event, 
   actionEventMapper: (event) => event}
);
```

2.- Removes the automatic loading of a datadog-configuration.json file by not passing a config file to FileBasedConfiguration(). The way in which this was done was flimsy, as it relied on jumping out of node_modules up to the root app folder. This worked most of the time but it caused an issue generating a require cycle that made 2 of our internal enums to be undefined at runtime (TrackingConsent and SdkVerbosity).

3.- Updates the FileBasedConfiguration tests to be more thorough and cover edge cases properly.

4.- Expands the example app showcasing different ways to provide a configuration object to the DatadogProvider, including FileBasedConfiguration schemes.

### Motivation

FileBasedConfiguration should consistently work across the board.

### Additional notes

This is a breaking change, as the option of automatically looking for a config.json file without having the user provide a specific path when creating a `FileBasedConfiguration` object will cease to work.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

### CODE REVIEW

File: benchmarks/README.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
+# Benchmark runner app for Datadog React Native SDK
+
+## Build and run
+
+1. From the root benchmark folder, run `yarn install`.
+2. From the iOS folder, run `pod install`.
+3. From the root folder, run `yarn ios` or `yarn android`.
+
+## Enable new architcture
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
## Enable new architecture
```

File: benchmarks/README.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+RCT_NEW_ARCH_ENABLED=1 pod install
+```
+
+### Android
+
+Set `newArchEnabled=true` in `benchmarks/android/gradle.properties`.
+
+## ENV config
+
+The `.env` config file contains configuration for the Datadog API, as well as a test scenario that you can launch when the app boots up. If it doesn't contain a test scenario, the app waits for a deeplink to trigger a test scenario instead.
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
The `.env` config file contains a configuration for the Datadog API, as well as a test scenario that you can launch when the app boots up. If it doesn't contain a test scenario, the app waits for a deep link to trigger a test scenario instead.
```

File: benchmarks/README.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+```
+    
+On iOS:
+``` 
+    xcodebuild -scheme alternate
+``` 
+
+## Set test scenarios using deeplinks
+
+You can trigger test scenarios by running a deeplink with a specific set of parameters to the benchmark app.
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
## Set test scenarios using deep links

You can trigger test scenarios by running a deep link with a specific set of parameters to the benchmark app.
```

File: benchmarks/README.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+## Run a test scenario
+
+Launch a specific test scenario with a particular configuration:
+- **Method**: `start`
+- **Parameters**:
+    - **scenario**: name of the test as defined on `Scenario` enum in `types/testConfig.ts`
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
    - **scenario**: name of the test as defined on the `Scenario` enum in `types/testConfig.ts`
```

File: benchmarks/README.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+```ts
+export enum Scenario {
+  Default = 'default',
+  NavigationExample = 'navigation',
+  (...)
+  NewScenario = 'newScenario'
+}
+```
+
+The string value (`'newScenario'`) is what must be passed using deeplink as the `scenario` parameter.
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
The string value (`'newScenario'`) is what must be passed using a deep link as the `scenario` parameter.
```

File: benchmarks/README.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+export default NewScenario;
+```
+
+### Register the Scenario in the App
+
+In `benchmarks/src/App.tsx`, add a new case to the scenario switch using the enum identifier you just created:
+
+```tsx
+case Scenario.NewScenario:
+  return <NewScenario testConfig={testConfig} />;
```
Comment (buraizu) [inline benchmarks/README.md:159]:
```suggestion
  return < NewScenario testConfig={testConfig} />;
```

File: benchmarks/README.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+```
+
+You can now trigger this scenario either through a deeplink or by setting `BENCH_SCENARIO` in the `.env` file.
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
You can then trigger this scenario through either a deep link or by setting `BENCH_SCENARIO` in the `.env` file.
```

File: benchmarks/README.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+xcrun simctl openurl booted "benchmark://stop"
+```
+
+### Android
+
+```
+adb shell am start -W -a android.intent.action.VIEW -d 'benchmark://stop' com.benchmarkrunner
+```
+
+## Adding a New Scenario to Benchmarks
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
## Add a new scenario to benchmarks
```

File: benchmarks/README.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+### Define the New Scenario
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
### Define the new scenario
```

File: benchmarks/README.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+The string value (`'newScenario'`) is what must be passed using deeplink as the `scenario` parameter.
+
+Alternatively, if you're using a `.env` file to set the scenario, add:
+
+```env
+BENCH_SCENARIO="newScenario"
+```
+
+### Create the Scenario Component
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
### Create the scenario component
```

File: benchmarks/README.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+import { NewScenarioProps } from "./types";
+
+function NewScenario(props: NewScenarioProps): React.JSX.Element {
+  (...)
+}
+
+export default NewScenario;
+```
+
+### Register the Scenario in the App
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
### Register the scenario in the app
```

File: benchmarks/README.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+In `benchmarks/src/App.tsx`, add a new case to the scenario switch using the enum identifier you just created:
+
+```tsx
+case Scenario.NewScenario:
+  return <NewScenario testConfig={testConfig} />;
+```
+
+You can now trigger this scenario either through a deeplink or by setting `BENCH_SCENARIO` in the `.env` file.
+
+## Instrumenting a Scenario
```
Comment (buraizu) [inline benchmarks/README.md]:
```suggestion
## Instrument a scenario
```

File: packages/react-native-babel-plugin/README.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
+# Babel Plugin for React Native
+
+The `@datadog/mobile-react-native-babel-plugin` enhances the Datadog React Native SDK by automatically enriching React components with contextual metadata. It helps improve the accuracy of features such as RUM event correlation, Session Replay, and UI tracking.
+
+## Setup
+
+**Note**: Make sure you’ve already integrated the [Datadog React Native SDK][1].
```
Comment (buraizu) [inline packages/react-native-babel-plugin/README.md]:
```suggestion
**Note**: Make sure you've already integrated the [Datadog React Native SDK][1].
```

File: packages/react-native-babel-plugin/README.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+Example configuration:
+
+```js
+module.exports = {
+  presets: ['module:@react-native/babel-preset'],
+  plugins: ['@datadog/mobile-react-native-babel-plugin'] // <-- Add here
+};
+```
+
+If you are currently using `actionNameAttribute` in your datadog SDK configuration, you'll need to also specify it here:
```
Comment (buraizu) [inline packages/react-native-babel-plugin/README.md]:
```suggestion
If you are using `actionNameAttribute` in your Datadog SDK configuration, you must also specify it here:
```

File: packages/react-native-babel-plugin/README.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+  plugins: [
+    [
+      '@datadog/mobile-react-native-babel-plugin',
+      {actionNameAttribute: 'custom-prop-value'},
+    ],
+  ],
+};
+```
+
+For more recent React Native versions this should be all that is needed. However, if you're on an older version and using Typescript in your project, you may need to install the preset `@babel/preset-typescript`.
```
Comment (buraizu) [inline packages/react-native-babel-plugin/README.md]:
```suggestion
**Note**: If you're on an older React Native version, and using Typescript in your project, you may need to install the preset `@babel/preset-typescript`.
```

File: packages/react-native-babel-plugin/README.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+npm install @babel/preset-typescript
+```
+
+To install with Yarn, run:
+
+```sh
+yarn add @babel/preset-typescript 
+```
+
+Then update your Babel configuration file like using the following example:
```
Comment (buraizu) [inline packages/react-native-babel-plugin/README.md:133]:
```suggestion
Then update your Babel configuration file as shown in the following example:
```

**Review by buraizu — COMMENTED** [review]
Just flagging some typos and suggesting a few other updates for consistency

**Comment by sbarrio** [comment]
> Just flagging some typos and suggesting a few other updates for consistency

@buraizu Hi, thanks for raising all those typos 🙇 

If you don't mind, I'll open a separate PR with your changes, as this specific PR is aimed toward a feature branch, and I think it's better to apply these corrections directly on develop.

EDIT: I've opened a PR with these fixes [here](https://github.com/DataDog/dd-sdk-reactnative/pull/974).

File: packages/core/src/sdk/FileBasedConfiguration/FileBasedConfiguration.ts [hunk 17]
```diff
@@ -56,44 +56,7 @@ export class FileBasedConfiguration extends DatadogProviderConfiguration {
...
 ): Record<string, any> => {
-    if (
-        userSpecifiedConfiguration === undefined ||
-        userSpecifiedConfiguration === null
-    ) {
-        try {
-            // This corresponds to a file located at the root of a RN project.
-            // /!\ We have to write the require this way as dynamic requires are not supported by Hermes.
-            // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
-            const jsonContent = require('../../../../../../datadog-configuration.json');
```
Comment (cdn34dd) [inline packages/core/src/sdk/FileBasedConfiguration/FileBasedConfiguration.ts:67]: So now we will no longer try to resolve the configuration file ourselves automatically ? We'll require it to be passed so we don't have to rely on that flaky path correct ?

**Review by cdn34dd — APPROVED** [review]

File: packages/core/src/sdk/FileBasedConfiguration/FileBasedConfiguration.ts (hunk 17 above)
Comment (sbarrio) [reply packages/core/src/sdk/FileBasedConfiguration/FileBasedConfiguration.ts:67]: Exactly, passing a specific path to the configuration file will be mandatory on v3 and we won't try to guess it on our own as it was prone to errors and ended up being flimsy.
